### ISSUE TITLE
[JENKINS-9913] Not obvious why some post-build tasks enforce serial behavior even when builds are concurrent

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#4542]
<p>We're experiencing an issue with concurrent builds where Jenkins appears to be associating separate builds (run on different machines) such that they won't be marked as completed until all jobs are completed.  For example, if we kick off 5 concurrent builds on 5 different nodes, builds 1-4 won't be marked as completed if build #5 is still running, even though builds 1-4 are finished.  I've seen a report of someone experiencing this issue elsewhere:</p>

<p><a href="http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1" class="external-link" target="_blank" rel="nofollow noopener">http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1</a></p>

<p>but a solution wasn't posted.  We do not have the batch plugin or the locks and latches plugin installed.  We've disabled all post-build processing and switched between different containers (Glassfish/Tomcat), but the problem persists.  I couldn't find an issue logged for this other than the aforementioned posting.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/pomvr.jpg" title="pomvr's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pomvr">pomvr</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-9913" target="_blank">Not obvious why some post-build tasks enforce serial behavior even when builds are concurrent</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>label(s)</b>: concurrent-build, junit
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2013-08-06T22:09:21+00:00
<li><b>votes</b>: 19
<li><b>watchers</b>: 34
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We're experiencing an issue with concurrent builds where Jenkins appears to be associating separate builds (run on different machines) such that they won't be marked as completed until all jobs are completed.  For example, if we kick off 5 concurrent builds on 5 different nodes, builds 1-4 won't be marked as completed if build #5 is still running, even though builds 1-4 are finished.  I've seen a report of someone experiencing this issue elsewhere:</p>

<p><a href="http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1" class="external-link" target="_blank" rel="nofollow noopener">http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1</a></p>

<p>but a solution wasn't posted.  We do not have the batch plugin or the locks and latches plugin installed.  We've disabled all post-build processing and switched between different containers (Glassfish/Tomcat), but the problem persists.  I couldn't find an issue logged for this other than the aforementioned posting.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>RedHat Enterprise Linux 4.8, Jenkins 1.414</code></li></ul>
<details><summary><i>2 attachments</i></summary>

- [jenkins1.jpeg](https://issues.jenkins.io/secure/attachment/22736/jenkins1.jpeg)
  > ![jenkins1.jpeg](https://issues.jenkins.io/secure/attachment/22736/jenkins1.jpeg)
- [jenkins2.jpeg](https://issues.jenkins.io/secure/attachment/22735/jenkins2.jpeg)
  > ![jenkins2.jpeg](https://issues.jenkins.io/secure/attachment/22735/jenkins2.jpeg)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ahawtho's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=158233&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-158233">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>We see the same thing - we have a test job that can be triggered with a few different upstream jobs to specify different sets of tests to run with different test parameters.  We use the parameterized trigger plugin to kick off multiple concurrent instances of the downstream job.  One of our "trigger" jobs specifies a set of long running performance tests (about 10 hours).  Another is a regression test suite, which only takes a few minutes.  We've seen instances at least twice where one of the regression tests will start after the long-running performance tests.  The short-lived regression test job will remain running until the performance test completes, which can be ten hours later.  The console log shows the regression test job as completed, having archived and recorded fingerprints, but it's still marked as running.  Furthermore, an attempt to stop the regression test job caused the long-running performance test job to stop.  Several regression tests would normally be queued during this time.  This seems to have been recent in an upgrade to 1.448.  I'm not sure what we were running immediately before that.</p>
</pre>
</details>

</details>
<p>We see the same thing - we have a test job that can be triggered with a few different upstream jobs to specify different sets of tests to run with different test parameters.  We use the parameterized trigger plugin to kick off multiple concurrent instances of the downstream job.  One of our "trigger" jobs specifies a set of long running performance tests (about 10 hours).  Another is a regression test suite, which only takes a few minutes.  We've seen instances at least twice where one of the regression tests will start after the long-running performance tests.  The short-lived regression test job will remain running until the performance test completes, which can be ten hours later.  The console log shows the regression test job as completed, having archived and recorded fingerprints, but it's still marked as running.  Furthermore, an attempt to stop the regression test job caused the long-running performance test job to stop.  Several regression tests would normally be queued during this time.  This seems to have been recent in an upgrade to 1.448.  I'm not sure what we were running immediately before that.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dannystaple.jpg" title="dannystaple's avatar" /> dannystaple's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=158438&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-158438">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>We were hoping that disabling the Junit post build task would stop this - it didn't do so. This has been killing us for some test jobs where we use a concurrent job to massively reduce duplication of configuration (and all the headaches that come with that).</p>
</pre>
</details>

</details>
<p>We were hoping that disabling the Junit post build task would stop this - it didn't do so. This has been killing us for some test jobs where we use a concurrent job to massively reduce duplication of configuration (and all the headaches that come with that).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>xavier_leprevost's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=160675&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-160675">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I have the same problem. 
I am using parameterized trigger where the parameters is a SVN branch.
I start :</p>
<ul class="alternate" type="square">
	<li>job build 1 from branch1</li>
	<li>job build 2 from branch2</li>
	<li>job build 3 from branch3</li>
	<li>job build 4 from branch4</li>
</ul>


<p>They are all running fine, but if job 2 or 3 or 4 finish before 1. They are really finish only when job 1 is finish. 
Then problem is the node resource not available then for starting a new job, and I am waiting the result of the build.</p>


</pre>
</details>

</details>
<p>I have the same problem. <br/>
I am using parameterized trigger where the parameters is a SVN branch.<br/>
I start :</p>
<ul class="alternate" type="square">
	<li>job build 1 from branch1</li>
	<li>job build 2 from branch2</li>
	<li>job build 3 from branch3</li>
	<li>job build 4 from branch4</li>
</ul>


<p>They are all running fine, but if job 2 or 3 or 4 finish before 1. They are really finish only when job 1 is finish. <br/>
Then problem is the node resource not available then for starting a new job, and I am waiting the result of the build.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=161568&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-161568">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Jenkins ver. 1.458
Still have problem: concurrent build not finished while last build ended. </p>
</pre>
</details>

</details>
<p>Jenkins ver. 1.458<br/>
Still have problem: concurrent build not finished while last build ended. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=161781&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-161781">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Any progress in resolve issue?</p>
</pre>
</details>

</details>
<p>Any progress in resolve issue?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=163332&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163332">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Do anybody read this ticket?
This bug make our work slow.
Help!</p>
</pre>
</details>

</details>
<p>Do anybody read this ticket?<br/>
This bug make our work slow.<br/>
Help!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=163653&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163653">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Created:    08/Jun/11
Today:      08/Jun/12
Still no progress...</p>
</pre>
</details>

</details>
<p>Created:    08/Jun/11<br/>
Today:      08/Jun/12<br/>
Still no progress...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sma.png" title="sma's avatar" /> sma's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=170997&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-170997">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Do we have a public target date for this bug to be fixed? 
It's getting kinda blocker for me.</p>
</pre>
</details>

</details>
<p>Do we have a public target date for this bug to be fixed? <br/>
It's getting kinda blocker for me.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=171000&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-171000">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I am afraid that no one is working on a the investigation of this bug.</p>
</pre>
</details>

</details>
<p>I am afraid that no one is working on a the investigation of this bug.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>xavier_leprevost's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=171001&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-171001">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>It would be nice to know if someone will investigate this problem.</p>
</pre>
</details>

</details>
<p>It would be nice to know if someone will investigate this problem.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>inbar_rose's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=172404&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-172404">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>same problem here. total blocker. task A starts, then task B starts. task B reaches the 'Recording test results' stage and hangs until task A finishes. after testing with simple timed builds with many plugins/options enabled/disabled concluded that junit is the problem. - found another issue like this here: <a href="https://issues.jenkins-ci.org/browse/JENKINS-10234" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-10234</a> </p>
</pre>
</details>

</details>
<p>same problem here. total blocker. task A starts, then task B starts. task B reaches the 'Recording test results' stage and hangs until task A finishes. after testing with simple timed builds with many plugins/options enabled/disabled concluded that junit is the problem. - found another issue like this here: <a href="https://issues.jenkins-ci.org/browse/JENKINS-10234" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-10234</a> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.jpg" title="kutzi's avatar" /> kutzi's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=172422&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-172422">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Similar to <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a> - for some causes even the same.</p>
</pre>
</details>

</details>
<p>Similar to <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a> - for some causes even the same.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.jpg" title="kutzi's avatar" /> kutzi's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=172423&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-172423">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>There are several cause described here. Some are JUnit archiving, which is specifically handled in <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a>. Some seem to be related to email notification.
I dare to say that in almost all cases this is a feature and not a bug as the logic for email notification and JUnit result archiving needs to wait until the previous builds are finished.</p>
</pre>
</details>

</details>
<p>There are several cause described here. Some are JUnit archiving, which is specifically handled in <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a>. Some seem to be related to email notification.<br/>
I dare to say that in almost all cases this is a feature and not a bug as the logic for email notification and JUnit result archiving needs to wait until the previous builds are finished.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=172425&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-172425">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>It is not a feature. All concurent builds execute in separated workspace and must  archive its artifacts independently. No need to wait others builds if current have finished.</p>
</pre>
</details>

</details>
<p>It is not a feature. All concurent builds execute in separated workspace and must  archive its artifacts independently. No need to wait others builds if current have finished.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.jpg" title="kutzi's avatar" /> kutzi's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=172427&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-172427">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Sergey, do you mean 'artifact archiving' or 'JUnit result archiving'. In the former case, you would be probably right, but I've seen no comment here that artifact archiving is blocking, too.
In the case of JUnit: yes it MUST block to calculate the diff (regressions et.al.) to the previous test results. So it is a feature.</p>
</pre>
</details>

</details>
<p>Sergey, do you mean 'artifact archiving' or 'JUnit result archiving'. In the former case, you would be probably right, but I've seen no comment here that artifact archiving is blocking, too.<br/>
In the case of JUnit: yes it MUST block to calculate the diff (regressions et.al.) to the previous test results. So it is a feature.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=172429&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-172429">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I mean all result of concurent job.
For example. I have Jenkins job that execute E2E tests on different QA servers. I started 10 jobs. One job running 5 hours. But others takes 1 hour. Then I can't see results of finished job and have to wait 4 hours to see it. We must have instrument to prevent this.</p>
</pre>
</details>

</details>
<p>I mean all result of concurent job.<br/>
For example. I have Jenkins job that execute E2E tests on different QA servers. I started 10 jobs. One job running 5 hours. But others takes 1 hour. Then I can't see results of finished job and have to wait 4 hours to see it. We must have instrument to prevent this.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svs57's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=179656&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-179656">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Thank you, 4 4. Now it's clear.
This will cause new problems, because my tests execute notification job in post-build step. This job analize what was upstream and send mail <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/>
I have to chage logic in notify job.</p>
</pre>
</details>

</details>
<p>Thank you, 4 4. Now it's clear.<br/>
This will cause new problems, because my tests execute notification job in post-build step. This job analize what was upstream and send mail <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
I have to chage logic in notify job.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dogfood.png" title="dogfood's avatar" /> dogfood's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=183137&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-183137">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/blue.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/2789/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #2789</a>
     <span class="error">&#91;FIXED JENKINS-9913&#93;</span> At least print a diagnostic to the build log if we are waiting on a checkpoint. (Revision eec307511c80112274d27f2a840d9f96cda784d3)</p>

<p>     Result = SUCCESS
Jesse Glick : <a href="https://github.com/jenkinsci/jenkins/commit/eec307511c80112274d27f2a840d9f96cda784d3" class="external-link" target="_blank" rel="nofollow noopener">eec307511c80112274d27f2a840d9f96cda784d3</a>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/tasks/junit/JUnitResultArchiver.java</li>
	<li>core/src/main/java/hudson/tasks/BuildStepMonitor.java</li>
	<li>core/src/main/resources/hudson/model/Messages.properties</li>
	<li>core/src/main/java/hudson/model/CheckPoint.java</li>
	<li>core/src/main/java/hudson/model/Run.java</li>
</ul>

</pre>
</details>

</details>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/blue.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/2789/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #2789</a><br/>
     <span class="error">&#91;FIXED JENKINS-9913&#93;</span> At least print a diagnostic to the build log if we are waiting on a checkpoint. (Revision eec307511c80112274d27f2a840d9f96cda784d3)</p>

<p>     Result = SUCCESS<br/>
Jesse Glick : <a href="https://github.com/jenkinsci/jenkins/commit/eec307511c80112274d27f2a840d9f96cda784d3" class="external-link" target="_blank" rel="nofollow noopener">eec307511c80112274d27f2a840d9f96cda784d3</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/tasks/junit/JUnitResultArchiver.java</li>
	<li>core/src/main/java/hudson/tasks/BuildStepMonitor.java</li>
	<li>core/src/main/resources/hudson/model/Messages.properties</li>
	<li>core/src/main/java/hudson/model/CheckPoint.java</li>
	<li>core/src/main/java/hudson/model/Run.java</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/surya548.jpg" title="surya548's avatar" /> surya548's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184112&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184112">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I am not sure how printing  to logs resolves this.  Can we reopen this.
Also can we make  Run.waitForCheckpoint an instance method instead of a static method.</p>
</pre>
</details>

</details>
<p>I am not sure how printing  to logs resolves this.  Can we reopen this.<br/>
Also can we make  Run.waitForCheckpoint an instance method instead of a static method.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184114&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184114">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>@surya548:</p>

<blockquote><p>I am not sure how printing to logs resolves this.</p></blockquote>

<p>As noted above, it just makes it clear why a given plugin is blocking the way it is. If there is a particular plugin which is blocking which you believe should not block, that should be filed in a separate issue.</p>

<blockquote><p>can we make Run.waitForCheckpoint an instance method instead of a static method</p></blockquote>

<p>I see no reason why that would be necessary, but this is probably not the place to discuss it anyway.</p>
</pre>
</details>

</details>
<p>@surya548:</p>

<blockquote><p>I am not sure how printing to logs resolves this.</p></blockquote>

<p>As noted above, it just makes it clear why a given plugin is blocking the way it is. If there is a particular plugin which is blocking which you believe should not block, that should be filed in a separate issue.</p>

<blockquote><p>can we make Run.waitForCheckpoint an instance method instead of a static method</p></blockquote>

<p>I see no reason why that would be necessary, but this is probably not the place to discuss it anyway.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/surya548.jpg" title="surya548's avatar" /> surya548's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184118&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184118">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>

<p>&gt; If there is a particular plugin which is blocking which you believe should not block</p>

<p>I dont want any plugin to block on checkpoint because it doesn't make sense to compare two arbitrary branches in the repo.</p>

<p>&gt; that should be filed in a separate issue. </p>

<p>I have so many plugins  that do this  here are just few example  example : junitarchiver, tap, checkstyle, findbugs , corbertura </p>

<p>I would like a global disable checkpointing option.</p>
</pre>
</details>

</details>

<p>> If there is a particular plugin which is blocking which you believe should not block</p>

<p>I dont want any plugin to block on checkpoint because it doesn't make sense to compare two arbitrary branches in the repo.</p>

<p>> that should be filed in a separate issue. </p>

<p>I have so many plugins  that do this  here are just few example  example : junitarchiver, tap, checkstyle, findbugs , corbertura </p>

<p>I would like a global disable checkpointing option.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184119&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184119">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<blockquote><p>it doesn't make sense to compare two arbitrary branches in the repo</p></blockquote>

<p>True, but this needs to be solved at a higher level in Jenkins, by creating a separate <tt>AbstractProject</tt> for each branch, so that each has a linear build history. There is some work TBA that accomplishes this.</p>

<blockquote><p>I would like a global disable checkpointing option.</p></blockquote>

<p>Sorry, this needs to be implemented on a per-plugin basis, since plugins are not necessarily written to behave gracefully when their checkpoint expectations are unmet.</p>
</pre>
</details>

</details>
<blockquote><p>it doesn't make sense to compare two arbitrary branches in the repo</p></blockquote>

<p>True, but this needs to be solved at a higher level in Jenkins, by creating a separate <tt>AbstractProject</tt> for each branch, so that each has a linear build history. There is some work TBA that accomplishes this.</p>

<blockquote><p>I would like a global disable checkpointing option.</p></blockquote>

<p>Sorry, this needs to be implemented on a per-plugin basis, since plugins are not necessarily written to behave gracefully when their checkpoint expectations are unmet.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/surya548.jpg" title="surya548's avatar" /> surya548's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184120&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184120">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>&gt; so that each has a linear build history.
We dont care about none of this linear history, checkpointing stuff if builds are hung for days deadlocked on checkpoints. 
There should be a way to turn this feature off. I think most people use git with lots of branches anyways these days and none of this is relevant/useful. </p>

<p>&gt;plugins are not necessarily written to behave gracefully when their checkpoint expectations are unmet</p>

<p>if we just noop  Run.waitForCheckpoint based on some global setting wouldn't plugins just assume there is nothing to compare against and behave gracefully? </p>
</pre>
</details>

</details>
<p>> so that each has a linear build history.<br/>
We dont care about none of this linear history, checkpointing stuff if builds are hung for days deadlocked on checkpoints. <br/>
There should be a way to turn this feature off. I think most people use git with lots of branches anyways these days and none of this is relevant/useful. </p>

<p>>plugins are not necessarily written to behave gracefully when their checkpoint expectations are unmet</p>

<p>if we just noop  Run.waitForCheckpoint based on some global setting wouldn't plugins just assume there is nothing to compare against and behave gracefully? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184125&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184125">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<blockquote><p>I think most people use git with lots of branches</p></blockquote>

<p>Exactly why separate branch projects are needed: so that each branch does <em>not</em> need to be marked concurrent-capable, avoiding wasted build and also avoiding waits on checkpoints as a corollary.</p>

<blockquote><p>if we just noop Run.waitForCheckpoint based on some global setting wouldn't plugins just assume there is nothing to compare against and behave gracefully?</p></blockquote>

<p>Depends on the plugin. Some may behave fine, but others may behave erratically because they are still looking up the “previous build” (which they assume to be complete based on their stated checkpoint semantics). Disabling the checkpoint is incompatible.</p>
</pre>
</details>

</details>
<blockquote><p>I think most people use git with lots of branches</p></blockquote>

<p>Exactly why separate branch projects are needed: so that each branch does <em>not</em> need to be marked concurrent-capable, avoiding wasted build and also avoiding waits on checkpoints as a corollary.</p>

<blockquote><p>if we just noop Run.waitForCheckpoint based on some global setting wouldn't plugins just assume there is nothing to compare against and behave gracefully?</p></blockquote>

<p>Depends on the plugin. Some may behave fine, but others may behave erratically because they are still looking up the “previous build” (which they assume to be complete based on their stated checkpoint semantics). Disabling the checkpoint is incompatible.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/surya548.jpg" title="surya548's avatar" /> surya548's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184188&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184188">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>&gt;Exactly why separate branch projects are needed.</p>

<p>I dont understand what you suggest that I do now. Our Jenkins Instance is basically unusable because of these useless checkpoints.
I am not sure if it is practical  for me to open bugs against 9 plugins( that i know of, after looking at source code for all the plugins we use) and ask them to remove checkpoints. </p>

<p>Why can't we get rid of (or atleast make it optional) a obsolete feature designed for SVN/CVS . 
As others have said in the comments above we would rather have stable Jenkins than some  optional feature than no-one cares about.</p>
</pre>
</details>

</details>
<p>>Exactly why separate branch projects are needed.</p>

<p>I dont understand what you suggest that I do now. Our Jenkins Instance is basically unusable because of these useless checkpoints.<br/>
I am not sure if it is practical  for me to open bugs against 9 plugins( that i know of, after looking at source code for all the plugins we use) and ask them to remove checkpoints. </p>

<p>Why can't we get rid of (or atleast make it optional) a obsolete feature designed for SVN/CVS . <br/>
As others have said in the comments above we would rather have stable Jenkins than some  optional feature than no-one cares about.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184198&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184198">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<blockquote><p>I am not sure if it is practical for me to open bugs against 9 plugins</p></blockquote>

<p>Why not?</p>

<blockquote><p>feature designed for SVN/CVS</p></blockquote>

<p>Applies equally to Git as to SVN; both support branches, and for any such SCM it is preferable to have one <tt>AbstractProject</tt> per branch so that each has a sensible linear history (as previously mentioned there is ongoing work in this regard). If you do that, and suppress parallel builds within a branch (i.e. may have concurrent builds only of distinct branches), then there is no further issue.</p>

<p>Now even within a branch it is sometimes desirable to permit parallel builds, when there is a premium on quick feedback over build-to-build comparisons; in that case plugins should be configurable to do no comparison to the previous build (i.e. you are intentionally waiving your right to information such as whether a given test case failure was a regression), and should not use checkpoints either.</p>

<blockquote><p>some optional feature</p></blockquote>

<p>The original behavior of build steps was to be unconditionally serialized. For compatibility, that default must remain. New plugins should consider their actual needs and override the relevant method to specify it (optionally introducing finer-grained checkpoints).</p>
</pre>
</details>

</details>
<blockquote><p>I am not sure if it is practical for me to open bugs against 9 plugins</p></blockquote>

<p>Why not?</p>

<blockquote><p>feature designed for SVN/CVS</p></blockquote>

<p>Applies equally to Git as to SVN; both support branches, and for any such SCM it is preferable to have one <tt>AbstractProject</tt> per branch so that each has a sensible linear history (as previously mentioned there is ongoing work in this regard). If you do that, and suppress parallel builds within a branch (i.e. may have concurrent builds only of distinct branches), then there is no further issue.</p>

<p>Now even within a branch it is sometimes desirable to permit parallel builds, when there is a premium on quick feedback over build-to-build comparisons; in that case plugins should be configurable to do no comparison to the previous build (i.e. you are intentionally waiving your right to information such as whether a given test case failure was a regression), and should not use checkpoints either.</p>

<blockquote><p>some optional feature</p></blockquote>

<p>The original behavior of build steps was to be unconditionally serialized. For compatibility, that default must remain. New plugins should consider their actual needs and override the relevant method to specify it (optionally introducing finer-grained checkpoints).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/surya548.jpg" title="surya548's avatar" /> surya548's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184200&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184200">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>&gt;Applies equally to Git as to SVN; both support branches.</p>

<p>There is a big big difference, svn branches are long lived and most git branches typically are short lived( pull requests for example) . And having linear history doesn't make any sense for a branch with only a few commits.
The same goes for calculating changesets, culprits etc. None of those features are relevant for short lived branches. </p>

</pre>
</details>

</details>
<p>>Applies equally to Git as to SVN; both support branches.</p>

<p>There is a big big difference, svn branches are long lived and most git branches typically are short lived( pull requests for example) . And having linear history doesn't make any sense for a branch with only a few commits.<br/>
The same goes for calculating changesets, culprits etc. None of those features are relevant for short lived branches. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184201&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184201">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Changelogs and the like are relevant if there is more than one commit in the branch, which is common in Git pull requests. And SVN branches need not be long lived since it is just as cheap to create and dispose of them as it is in Git. (CVS is a different matter of course.) Whether you care about linear build history depends on your workflow, not the SCM per se.</p>
</pre>
</details>

</details>
<p>Changelogs and the like are relevant if there is more than one commit in the branch, which is common in Git pull requests. And SVN branches need not be long lived since it is just as cheap to create and dispose of them as it is in Git. (CVS is a different matter of course.) Whether you care about linear build history depends on your workflow, not the SCM per se.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184210&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184210">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Around half of those have more than one commit, as I said.</p>
</pre>
</details>

</details>
<p>Around half of those have more than one commit, as I said.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/surya548.jpg" title="surya548's avatar" /> surya548's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184243&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184243">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I give up, you win man. thanks for your help.</p>
</pre>
</details>

</details>
<p>I give up, you win man. thanks for your help.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick's <a href="https://issues.jenkins.io/browse/JENKINS-9913?focusedId=184293&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184293">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I am trying to get a patch through in <a href="https://issues.jenkins.io/browse/JENKINS-16376" title="BuildStepMonitor.BUILD makes concurrent builds wait, could be changed to BuildStepMonitor.NONE?" class="issue-link" data-issue-key="JENKINS-16376"><del>JENKINS-16376</del></a>; please push for changes in other affected plugins.</p>
</pre>
</details>

</details>
<p>I am trying to get a patch through in <a href="https://issues.jenkins.io/browse/JENKINS-16376" title="BuildStepMonitor.BUILD makes concurrent builds wait, could be changed to BuildStepMonitor.NONE?" class="issue-link" data-issue-key="JENKINS-16376"><del>JENKINS-16376</del></a>; please push for changes in other affected plugins.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-15/issues?q=in%3Atitle%20JENKINS-16376">JENKINS-16376</a>]</i>
